lib/sdam/server_description.js

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+// An enumeration of server types we know about
+const ServerType = {
+  Standalone: 'Standalone',
+  Mongos: 'Mongos',
+  PossiblePrimary: 'PossiblePrimary',
+  RSPrimary: 'RSPrimary',
+  RSSecondary: 'RSSecondary',
+  RSArbiter: 'RSArbiter',
+  RSOther: 'RSOther',
+  RSGhost: 'RSGhost',
+  Unknown: 'Unknown'
+};
+
+/**
+ * The client's view of a single server, based on the most recent ismaster outcome.
+ *
+ * Internal type, not meant to be directly instantiated
+ */
+class ServerDescription {
+  /**
+   * Create a ServerDescription
+   * @param {String} address The address of the server
+   * @param {Object} [ismaster] An optional ismaster response for this server
+   * @param {Object} [options] Optional settings
+   * @param {Number} [options.roundTripTime] The round trip time to ping this server (in ms)
+   */
+  constructor(address, ismaster, options) {
+    options = options || {};
+
+    this.address = address;
+    this.error = null;
+    this.roundTripTime = options.roundTripTime || 0;
+    this.lastWriteDate = ismaster && ismaster.lastWrite ? ismaster.lastWrite.lastWriteDate : null;
+    this.opTime = ismaster && ismaster.lastWrite ? ismaster.lastWrite.opTime : null;
+    this.type = parseServerType(ismaster);
+    this.minWireVersion = (ismaster && ismaster.minWireVersion) || 0;
```

Comment:
I feel like a lot of these could be accomplished with an:

```js
Object.assign(this, { ...defaultValues }, ismaster);
```

Reply:
It could for sure, but then we would be vulnerable to potentially changing fields in future `ismaster`s. This way we are very explicit about what we are plucking from the `ismaster`, and the definition of this type only changes when we say it changes.

Reply:
in that case, could we do an 

```js
ismaster = Object.assign({ ...defaultValues }, ismaster);

this.lastWriteDate = ismaster.lastWriteDate;
...
```

Reply:
sure